File: Design Patterns/AbsctractFactory Pattern/abstractFactory.py

# In this example, the GUIFactory is an abstract factory that declares methods for creating button and checkbox products.
# There are two concrete factories: WindowsFactory and MacOSFactory, each of which provides implementations for creating their respective button and checkbox products.

# The abstract product classes Button and Checkbox define the methods that the concrete products (WindowsButton, MacOSButton, WindowsCheckbox, MacOSCheckbox) must implement.
# The application function demonstrates how the client code can use the abstract factory to create GUI elements specific to the chosen operating system.
# The client code takes user input to determine the desired operating system type and then creates the appropriate factory to produce the corresponding GUI elements.


# Abstract Factory
class GUIFactory:
    def create_button(self):
        pass

    def create_checkbox(self):
        pass

# Concrete Factory for Windows
class WindowsFactory(GUIFactory):
    def create_button(self):
        return WindowsButton()

    def create_checkbox(self):
        return WindowsCheckbox()

# Concrete Factory for macOS
class MacOSFactory(GUIFactory):
    def create_button(self):
        return MacOSButton()

    def create_checkbox(self):
        return MacOSCheckbox()

# Abstract Product: Button
class Button:
    def paint(self):
        pass

# Concrete Product: Windows Button
class WindowsButton(Button):
    def paint(self):
        return "Render a Windows button"

# Concrete Product: macOS Button
class MacOSButton(Button):
    def paint(self):
        return "Render a macOS button"

# Abstract Product: Checkbox
class Checkbox:
    def paint(self):
        pass

# Concrete Product: Windows Checkbox
class WindowsCheckbox(Checkbox):
    def paint(self):
        return "Render a Windows checkbox"

# Concrete Product: macOS Checkbox
class MacOSCheckbox(Checkbox):
    def paint(self):
        return "Render a macOS checkbox"

def application(factory):
    button = factory.create_button()
    checkbox = factory.create_checkbox()

    print(button.paint())
    print(checkbox.paint())

# Client code
if __name__ == "__main__":
    os_type = input("Enter the OS type (windows/mac): ")

    if os_type.lower() == "windows":
        factory = WindowsFactory()
    elif os_type.lower() == "mac":
        factory = MacOSFactory()
    else:
        raise ValueError("Invalid OS type")

    application(factory)

File: Design Patterns/AbstractFactory Pattern/abstractFactory.ts
import readline from "readline";

// Abstract Factory
abstract class GUIFactory {
  abstract createButton(): Button;
  abstract createCheckbox(): Checkbox;
}

// Concrete Factory for Windows
class WindowsFactory extends GUIFactory {
  createButton(): Button {
    return new WindowsButton();
  }

  createCheckbox(): Checkbox {
    return new WindowsCheckbox();
  }
}

// Concrete Factory for macOS
class MacOSFactory extends GUIFactory {
  createButton(): Button {
    return new MacOSButton();
  }

  createCheckbox(): Checkbox {
    return new MacOSCheckbox();
  }
}


// Abstract Product: Button
abstract class Button {
  abstract paint(): string;
}

// Concrete Product: Windows Button
class WindowsButton extends Button {
  paint(): string {
    return "Render a Windows button";
  }
}

// Concrete Product: macOS Button
class MacOSButton extends Button {
  paint(): string {
    return "Render a macOS button";
  }
}

// Abstract Product: Checkbox
abstract class Checkbox {
  abstract paint(): string;
}

// Concrete Product: Windows Checkbox
class WindowsCheckbox extends Checkbox {
  paint(): string {
    return "Render a Windows checkbox";
  }
}

// Concrete Product: macOS Checkbox
class MacOSCheckbox extends Checkbox {
  paint(): string {
    return "Render a macOS checkbox";
  }
}

function application(factory: GUIFactory) {
  const button = factory.createButton();
  const checkbox = factory.createCheckbox();

  console.log(button.paint());
  console.log(checkbox.paint());
}

// Client code

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

rl.question("Enter the OS type (windows/mac): ", (osType) => {
  rl.close();

  if (osType?.toLowerCase() === "windows") {
    const factory = new WindowsFactory();
    application(factory);
  } else if (osType?.toLowerCase() === "mac") {
    const factory = new MacOSFactory();
    application(factory);
  } else {
    throw new Error("Invalid OS type");
  }
});
